url7f729337e3_timmolter_Sundial_tgz-pJ8-org_knowm_sundial_SampleRunJ8/src/org/quartz/triggers/CronExpression.java

```diff
@@ -0,0 +1,1571 @@
+    return getTimeAfter(date);
+  }
+
+  /**
+   * Returns the next date/time <I>after</I> the given date/time which does <I>not</I> satisfy the expression
+   *
+   * @param date the date/time at which to begin the search for the next invalid date/time
+   * @return the next valid date/time
+   */
+  public Date getNextInvalidTimeAfter(Date date) {
+    long difference = 1000;
+
+    //move back to the nearest second so differences will be accurate
+    Calendar adjustCal = Calendar.getInstance(getTimeZone());
+    adjustCal.setTime(date);
+    adjustCal.set(Calendar.MILLISECOND, 0);
+    Date lastDate = adjustCal.getTime();
+
+    Date newDate;
+
+    //FUTURE_TODO: (QUARTZ-481) IMPROVE THIS! The following is a BAD solution to this problem. Performance will be very bad here, depending on the cron expression. It is, however A solution.
+
+    //keep getting the next included time until it's farther than one second
+    // apart. At that point, lastDate is the last valid fire time. We return
+    // the second immediately following it.
+    while (difference == 1000) {
+      newDate = getTimeAfter(lastDate);
+      if (newDate == null) {
+        break;
+      }
+
+      difference = newDate.getTime() - lastDate.getTime();
+
+      if (difference == 1000) {
+        lastDate = newDate;
+      }
+    }
+
+    return new Date(lastDate.getTime() + 1000);
+  }
+
+  /**
+   * Returns the time zone for which this <code>CronExpression</code> will be resolved.
+   */
+  public TimeZone getTimeZone() {
+    if (timeZone == null) {
+      timeZone = TimeZone.getDefault();
+    }
+
+    return timeZone;
+  }
+
+  /**
+   * Sets the time zone for which this <code>CronExpression</code> will be resolved.
+   */
+  public void setTimeZone(TimeZone timeZone) {
+    this.timeZone = timeZone;
+  }
+
+  /**
+   * Returns the string representation of the <CODE>CronExpression</CODE>
+   *
+   * @return a string representation of the <CODE>CronExpression</CODE>
+   */
+  @Override
+  public String toString() {
+    return cronExpression;
+  }
+
+  /**
+   * Indicates whether the specified cron expression can be parsed into a valid cron expression
+   *
+   * @param cronExpression the expression to evaluate
+   * @return a boolean indicating whether the given expression is a valid cron expression
+   */
+  public static boolean isValidExpression(String cronExpression) {
+
+    try {
+      new CronExpression(cronExpression);
+    } catch (ParseException pe) {
+      return false;
+    }
+
+    return true;
+  }
+
+  public static void validateExpression(String cronExpression) throws ParseException {
+
+    new CronExpression(cronExpression);
+  }
+
+  ////////////////////////////////////////////////////////////////////////////
+  //
+  // Expression Parsing Functions
+  //
+  ////////////////////////////////////////////////////////////////////////////
+
+  protected void buildExpression(String expression) throws ParseException {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 30. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.